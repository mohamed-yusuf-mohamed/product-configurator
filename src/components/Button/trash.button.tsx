// const StyledButton = styled.button<Props.Button>`
//   all: unset;
//   border: ${({ isSelected, theme }) =>
//     isSelected
//       ? theme.option.button.selected.border
//       : theme.option.button.unselected.border};
//   padding: ${({ theme }) => theme.option.button.padding};
//   cursor: pointer;
//   border-radius: ${({ theme }) => theme.option.button.border.radius};
// `


// const Button = (props: Props.Button) => {
  // const {isSelected, } = props
//   return <button {...props}>{children}</button>
// }
// export default Button

export{}
