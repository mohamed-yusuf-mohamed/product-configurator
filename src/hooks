import { TypedUseSelectorHook, useDispatch as _useDispatch, useSelector as _useSelector } from 'react-redux';
import type { State } from './store';
import { ThunkDispatch } from './store';

export const useDispatch = () => _useDispatch<ThunkDispatch>();
export const useSelector: TypedUseSelectorHook<State> = _useSelector;
