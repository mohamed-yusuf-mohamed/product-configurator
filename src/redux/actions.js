export const updateSelection =
  (options): Thunk =>
  (dispatch, getState) => {
    const { products } = getState()

    const select = () => {
      const hasOptions = (product) => {
        for (const type in options) {
          if (product.attributes[type].value !== options[type].value) {
            return false
          }
        }
        return true
      }

      for (const id in products) {
        if (!hasOptions(products[id])) continue
        return products[id]
      }
    }

    const product = select()

    dispatch({
      type: "UPDATE_SELECTION",
      payload: product,
    })
  }

export const initialiseState = (initialState) => ({
  type: "INITIALISE",
  payload: initialState,
})
