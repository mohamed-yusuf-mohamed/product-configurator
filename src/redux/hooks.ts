import { TypedUseSelectorHook, useDispatch as useDispatchFn, useSelector as useSelectorFn } from 'react-redux';
import type { State } from './store';
import { ThunkDispatch } from './store';

export const useDispatch = () => useDispatchFn<ThunkDispatch>();
export const useSelector: TypedUseSelectorHook<State> = useSelectorFn;
