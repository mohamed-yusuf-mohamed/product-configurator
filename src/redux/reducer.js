const reducer = (state, action) => {
  switch (action.type) {
    case "INITIALISE": {
      return action.payload
    }
    case "UPDATE_SELECTION": {
      return {
        ...state,
        selected: action.payload,
      }
    }
    default:
      return state
  }
}
export default reducer
