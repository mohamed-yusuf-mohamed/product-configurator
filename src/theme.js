import { parseHSL } from "./utils";
// Colours are defined in HSL format. https://css-tricks.com/yay-for-hsla

const theme = {
  // configurator: {
  title: {
    colour: "hsl(0, 0%, 12%)",
  },
  description: {
    colour: "hsl(0, 0%, 21%)",
  },
  font: "https://fonts.googleapis.com/css2?family=Roboto&display=swap",
  option: {
    button: {
      selected: {
        border: `1px solid hsl(164, 100%, 30%)`,
      },
      unselected: {
        border: "1px solid hsl(0, 0%, 21%)",
      },
      padding: "20px",
      border: {
        radius: "2rem",
      },
    },
  },
};

export default theme;
