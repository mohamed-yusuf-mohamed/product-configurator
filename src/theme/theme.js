const theme = {
  title: {
    colour: "hsl(0, 0%, 12%)",
  },
  description: {
    colour: "hsl(0, 0%, 21%)",
  },
  font: "https://fonts.googleapis.com/css2?family=Roboto&display=swap",
  option: {
    button: {
      selected: {
        border: `3px solid hsl(164, 100%, 30%)`,
      },
      unselected: {
        border: "2px solid hsl(0, 0%, 80%)",
      },
      padding: ".7rem",
      border: {
        radius: ".5rem",
      },
    },
  },
  summary: {
    table: {
      border: "1px solid hsl(0, 0%, 80%)",
      cell: {
        padding: "1rem",
      },
    },
  },
}

export default theme
