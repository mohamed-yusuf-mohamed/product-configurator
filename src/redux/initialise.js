import {initialiseData} from "./actions"
import {store} from "./store"
import api from "../api"
import {generateProducts, generateOptions} from "./initialise/generators"
const initialise = async () => {
  const data = await api.getProducts()
  const products = generateProducts(data)
  const options = generateOptions(data)
  const defaultSelection = products[1]

  store.dispatch(initialiseData({
    products,
    options,
    selected: defaultSelection
  }))
}
export default initialise