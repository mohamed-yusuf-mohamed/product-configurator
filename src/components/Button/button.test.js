import theme from "../../theme"
import { render} from "@testing-library/react"
import Button from "./index"
// import props from "./mocks/props.json"
import {useDispatch} from "../../redux/hooks"
// import {updateSelection} from "../../redux/actions"
import { ThemeProvider } from "styled-components"


jest.mock("../../redux/hooks", () => ({
  useDispatch: () => {}
}))
// jest.mock(updateSelection)

describe("button component", () => {
  const { getByTestId } = render(
    <ThemeProvider theme={theme}>
      <Button data-testid="option-button-selected" isSelected={true} type="cover-colour" />
      <Button data-testid="option-button-unselected" isSelected={false} type="cover-colour" />
    </ThemeProvider>
  )

  const component = {
    selected: getByTestId("option-button-selected"),
    unselected: getByTestId("option-button-unselected"),
  }

  it("renders: selected", () => {
    expect(component.selected).toMatchSnapshot()
  })

  it("renders: unselected", () => {
    expect(component.unselected).toMatchSnapshot()
  })
})
