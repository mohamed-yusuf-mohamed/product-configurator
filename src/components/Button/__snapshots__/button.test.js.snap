// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button component renders: selected 1`] = `
<button
  class="sc-bcXHqe fRBfvX"
  data-testid="option-button-selected"
  type="cover-colour"
/>
`;

exports[`button component renders: unselected 1`] = `
<button
  class="sc-bcXHqe dUuebz"
  data-testid="option-button-unselected"
  type="cover-colour"
/>
`;
